lms/djangoapps/shoppingcart/pdf.py

```diff
@@ -0,0 +1,448 @@
+"""
+Template for PDF Receipt/Invoice Generation
+"""
+from PIL import Image
+from reportlab.lib import colors
+from django.conf import settings
+from django.utils.translation import ugettext as _
+from reportlab.pdfgen.canvas import Canvas
+from reportlab.lib.pagesizes import letter
+from reportlab.lib.units import mm
+from reportlab.lib.styles import getSampleStyleSheet
+from reportlab.platypus import Paragraph
+from reportlab.platypus.tables import Table, TableStyle
+from microsite_configuration import microsite
+from xmodule.modulestore.django import ModuleI18nService
+
+
+class NumberedCanvas(Canvas):  # pylint: disable=abstract-method
+    """
+    Canvas child class with auto page-numbering.
+    """
+    def __init__(self, *args, **kwargs):
+        """
+            __init__
+        """
+        Canvas.__init__(self, *args, **kwargs)
+        self._saved_page_states = []
+
+    def insert_page_break(self):
+        """
+        Starts a new page.
+        """
+        self._saved_page_states.append(dict(self.__dict__))
+        self._startPage()
+
+    def current_page_count(self):
+        """
+        Returns the page count in the current pdf document.
+        """
+        return len(self._saved_page_states) + 1
+
+    def save(self):
+        """
+            Adds page numbering to each page (page x of y)
+        """
+        num_pages = len(self._saved_page_states)
+        for state in self._saved_page_states:
+            self.__dict__.update(state)
+            if num_pages > 1:
+                self.draw_page_number(num_pages)
+            Canvas.showPage(self)
+        Canvas.save(self)
+
+    def draw_page_number(self, page_count):
+        """
+        Draws the String "Page x of y" at the bottom right of the document.
+        """
+        self.setFontSize(7)
+        self.drawRightString(200 * mm, 12 * mm,
+                             _("Page %(page_number)d of %(page_count)d") % {"page_number": self._pageNumber,
+                                                                            "page_count": page_count})
+
+
+class PDFInvoice(object):
+    """
+    PDF Generation Class
+    """
+    def __init__(self, context):
+        """
+        The context is expected to be a dict having the following items:
+
+        items_data - A list having the following items for each row.
+            item_description - String
+            quantity - Integer
+            list_price - float
+            discount - float
+            item_total - float
```

Comment:
Since these parameters are required, I'd prefer that they be positional arguments.